buildcache/directory_cache.go

```diff
@@ -0,0 +1,100 @@
+// This file is part of arduino-cli.
+//
+// Copyright 2020 ARDUINO SA (http://www.arduino.cc/)
+//
+// This software is released under the GNU General Public License version 3,
+// which covers the main part of arduino-cli.
+// The terms of this license can be found at:
+// https://www.gnu.org/licenses/gpl-3.0.en.html
+//
+// You can be released from the requirements of the above licenses by purchasing
+// a commercial license. Buying such a license is mandatory if you want to
+// modify or otherwise use the software for commercial activities involving the
+// Arduino software without disclosing the source code of your own applications.
+// To purchase a commercial license, send an email to license@arduino.cc.
+
+package buildcache
+
+import (
+	"time"
+
+	"github.com/arduino/go-paths-helper"
+	"github.com/pkg/errors"
+	"github.com/sirupsen/logrus"
+)
+
+const (
+	lastUsedFileName = ".last-used"
+)
+
+type directoryCache struct {
+	baseDir *paths.Path
+	ttl     time.Duration
+}
+
+func (dc *directoryCache) isExpired(key string) time.Duration {
+	modTime, _ := dc.modTime(key)
+	return dc.ttl - time.Since(modTime)
+}
+
+func (dc *directoryCache) modTime(key string) (time.Time, error) {
+	fileInfo, err := dc.baseDir.Join(key, lastUsedFileName).Stat()
+	if err != nil {
+		// folders with a missing last used file are not purged
+		return time.Now().Add(time.Minute), err
+	}
+
+	return fileInfo.ModTime(), nil
+}
+
+func (dc *directoryCache) GetOrCreate(key string, value time.Time) (time.Time, error) {
+	existing := true
+	modTime, err := dc.modTime(key)
+	if err != nil {
+		existing = false
+	}
+
+	subDir := dc.baseDir.Join(key)
+	err = subDir.MkdirAll()
+	if err != nil {
+		return modTime, err
+	}
+	err = subDir.Join(lastUsedFileName).WriteFile([]byte{})
+	if err != nil || existing {
+		return modTime, err
+	}
+	return time.Now(), nil
+}
+
+func (dc *directoryCache) Purge() error {
+	files, err := dc.baseDir.ReadDir()
+	if err != nil {
+		return err
+	}
+	for _, file := range files {
+		if file.IsDir() {
+			dc.removeIfExpired(file)
+		}
+	}
+	return nil
+}
+
+func (dc *directoryCache) removeIfExpired(dir *paths.Path) {
+	lifeExpectancy := dc.isExpired(dir.Base())
+	if lifeExpectancy > 0 {
+		return
+	}
+	logrus.Tracef(`Purging cache directory "%s". Expired by %s\n`, dir, lifeExpectancy)
+	err := dir.RemoveAll()
+
+	if err != nil {
+		logrus.Tracef(`Error while pruning cache directory "%s".\n%s\n`, dir, errors.WithStack(err))
```

Comment:
```suggestion
	logrus.Tracef(`Purging cache directory "%s". Expired by %s`, dir, lifeExpectancy)
	err := dir.RemoveAll()

	if err != nil {
		logrus.Tracef(`Error while pruning cache directory "%s": %s`, dir, errors.WithStack(err))
```

- Line breaks are not needed at the end of log entries because each log entry is already a line.
- Line breaks don't work within log entries. You can use fields instead: https://github.com/sirupsen/logrus#fields